gobblin-modules/gobblin-orc/src/main/java/org/apache/gobblin/writer/GobblinBaseOrcWriter.java

```diff
@@ -68,29 +97,50 @@ public GobblinBaseOrcWriter(FsDataWriterBuilder<S, D> builder, State properties)
     this.inputSchema = builder.getSchema();
     this.typeDescription = getOrcSchema();
     this.valueWriter = getOrcValueWriter(typeDescription, this.inputSchema, properties);
-    this.batchSize = properties.getPropAsInt(ORC_WRITER_BATCH_SIZE, DEFAULT_ORC_WRITER_BATCH_SIZE);
+    this.selfTuningWriter = properties.getPropAsBoolean(ORC_WRITER_AUTO_SELFTUNE_ENABLED, false);
+    this.batchSize = this.selfTuningWriter ? properties.getPropAsInt(ORC_WRITER_BATCH_SIZE, DEFAULT_ORC_WRITER_BATCH_SIZE) : DEFAULT_ORC_WRITER_BATCH_SIZE;
     this.rowBatchPool = RowBatchPool.instance(properties);
     this.enableRowBatchPool = properties.getPropAsBoolean(RowBatchPool.ENABLE_ROW_BATCH_POOL, false);
+    this.selfTuneFrequency = properties.getPropAsInt(ORC_WRITER_AUTO_SELFTUNE_FREQUENCY, DEFAULT_ORC_AUTO_SELFTUNE_FREQUENCY);
+    // First tune is at 5
     this.rowBatch = enableRowBatchPool ? rowBatchPool.getRowBatch(typeDescription, batchSize) : typeDescription.createRowBatch(batchSize);
-    log.info("Created ORC writer, batch size: {}, {}: {}",
-            batchSize, OrcConf.ROWS_BETWEEN_CHECKS.getAttribute(),
-            properties.getProp(
-                    OrcConf.ROWS_BETWEEN_CHECKS.getAttribute(),
-                    OrcConf.ROWS_BETWEEN_CHECKS.getDefaultValue().toString()));
-
+    this.converterMemoryManager = new OrcConverterMemoryManager(this.rowBatch);
     // Create file-writer
-    Configuration conf = new Configuration();
+    this.writerConfig = new Configuration();
     // Populate job Configurations into Conf as well so that configurations related to ORC writer can be tuned easily.
     for (Object key : properties.getProperties().keySet()) {
-      conf.set((String) key, properties.getProp((String) key));
+      this.writerConfig.set((String) key, properties.getProp((String) key));
     }
-
-    OrcFile.WriterOptions options = OrcFile.writerOptions(properties.getProperties(), conf);
+    OrcFile.WriterOptions options = OrcFile.writerOptions(properties.getProperties(), this.writerConfig);
     options.setSchema(typeDescription);
 
-    // For buffer-writer, flush has to be executed before close so it is better we maintain the life-cycle of fileWriter
-    // instead of delegating it to closer object in FsDataWriter.
-    this.orcFileWriter = OrcFile.createWriter(this.stagingFile, options);
+    // Get the amount of allocated and future space available
+    this.availableMemory = Runtime.getRuntime().maxMemory() - (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory());
+    log.info("Available memory for ORC writer: {}", this.availableMemory);
+
+    if (this.selfTuningWriter) {
+      if (properties.contains(ORC_WRITER_ESTIMATED_RECORD_SIZE)) {
+        log.info("Using previously stored properties, ORC Estimated Record size is : {}, estimated bytes converter allocated is : {}",
```

Comment:
move to log after line 127 to reuse estimatedRecordSize and estimatedBytesAllocatedConverterMemory?